thunder/core/transforms.py

```diff
@@ -3736,9 +3737,99 @@ def backward_fn(saved_for_backward, cotangents):
+
+    saved_for_bw = get_saved_for_backward_tensors(fwd_trace)
+    fwd_trace_args = {variableify(j) for j in fwd_trace.args}
+    old_saved_for_bwd = {variableify(j) for j in saved_for_bw}
+
+    all_rematerializable = old_saved_for_bwd - fwd_trace_args
+
+    # Here will come the selection logic for what to materialize.
+    from collections import Counter
+
+    tensor_counter = Counter({v: unvariableify(v).numel * unvariableify(v).dtype.bytes for v in all_rematerializable})
+    print(f"DEBUG: out of {len(saved_for_bw)} sfb, {len(all_rematerializable)} are recomputable")
+    print(f"DEBUG: can save up to {tensor_counter.total()*1e-6:.3f} MB")
+    top_one_third = len(saved_for_bw) // 3
+    print(f"DEBUG: remat {top_one_third} tensors (top one third)")
+
+    # Select the tensors that will be rematerialized
+    rematerializable = {k for k, v in tensor_counter.most_common(top_one_third)}
+
+    producers = find_producer_symbols(fwd_trace, tuple(unvariableify(i) for i in rematerializable), fwd_trace.args)
+
+    required_fw_args = fwd_trace_args & old_saved_for_bwd
+    recomputed_tensors_from_producers = set()
+    for prod in producers:
+        for prod_arg in prod.flat_args:
+            prod_arg = variableify(prod_arg)
+            if prod_arg in fwd_trace_args:
+                required_fw_args.add(prod_arg)
+        for prod_out in prod.flat_outs:
+            recomputed_tensors_from_producers.add(variableify(prod_out))
+
+    required_saved_for_bwd = all_rematerializable - rematerializable - recomputed_tensors_from_producers
+    new_saved_for_backward = tuple(unvariableify(i) for i in required_fw_args | required_saved_for_bwd)
+
+    new_fwd_trace = from_trace(fwd_trace)
+    new_fwd_trace.bound_symbols = fwd_trace.bound_symbols.copy()
+    new_return_args = (fwd_trace.output[0], (new_saved_for_backward, fwd_trace.output[1][1]))
+    new_fwd_trace.bound_symbols[-1] = prims.python_return.bind(new_return_args, output=())
+
+    new_bwd_trace = from_trace(bwd_trace)
+
+    with tracectx(new_bwd_trace):
+        unpack_args = (CollectionProxy(new_saved_for_backward, name="C0"), len(new_saved_for_backward))
+
+    assert bwd_trace.bound_symbols[4].sym.id == prims.PrimIDs.UNPACK_SEQUENCE
+    assert bwd_trace.bound_symbols[4].args[0].name == "C0"
+    assert bwd_trace.bound_symbols[5].sym.id == prims.PrimIDs.UNPACK_SEQUENCE
+    assert bwd_trace.bound_symbols[5].args[0].name == "C1"
+
+    for idx, bsym in enumerate(bwd_trace.bound_symbols):
+        if idx == 4:
+            new_unpack = prims.unpack_sequence.bind(*unpack_args, output=new_saved_for_backward)
+            new_bwd_trace.bound_symbols.append(new_unpack)
+        elif idx == 6:
+            new_bwd_trace.bound_symbols.extend(producers)
+            new_bwd_trace.bound_symbols.append(bsym)
+        else:
+            new_bwd_trace.bound_symbols.append(bsym)
+
+    new_bwd_trace.args = [(new_saved_for_backward, fwd_trace.output[1][1]), *bwd_trace.args[1:]]
```

Comment:
How difficult was it to figure out the correct "preamble" of the backward trace? Is there any learning that you could add as a comment for future readers about why it's done as it is?